File: game_files/game_bot.py
import random
import math

class GameBot:
    def __init__(self, board_size=8):
        self.board_size = board_size
        self.board = None
        self.current_player = 1  # 1 for player, -1 for bot
        self.initialize_board()

    def initialize_board(self):
        self.board = [[0 for _ in range(self.board_size)] for _ in range(self.board_size)]

    def get_moves(self):
        return [(i, j) for i in range(self.board_size) for j in range(self.board_size) if self.board[i][j] == 0]

    def make_move(self, move):
        i, j = move
        if self.board[i][j] == 0:
            self.board[i][j] = self.current_player
            self.current_player *= -1  # Switch player
            return True
        return False

    def undo_move(self, move):
        i, j = move
        self.board[i][j] = 0
        self.current_player *= -1  # Switch player back

    def evaluate(self):
        return sum(sum(row) for row in self.board)

    def alpha_beta_pruning(self, depth, alpha, beta, maximizing_player):
        if depth == 0:
            return self.evaluate()

        moves = self.get_moves()

        if maximizing_player:
            max_eval = -math.inf
            for move in moves:
                self.make_move(move)
                eval = self.alpha_beta_pruning(depth - 1, alpha, beta, False)
                self.undo_move(move)
                max_eval = max(max_eval, eval)
                alpha = max(alpha, eval)
                if beta <= alpha:
                    break
            return max_eval
        else:
            min_eval = math.inf
            for move in moves:
                self.make_move(move)
                eval = self.alpha_beta_pruning(depth - 1, alpha, beta, True)
                self.undo_move(move)
                min_eval = min(min_eval, eval)
                beta = min(beta, eval)
                if beta <= alpha:
                    break
            return min_eval

    def get_best_move(self, depth=3):
        best_move = None
        best_eval = -math.inf
        alpha = -math.inf
        beta = math.inf
        moves = self.get_moves()

        for move in moves:
            self.make_move(move)
            eval = self.alpha_beta_pruning(depth - 1, alpha, beta, False)
            self.undo_move(move)

            if eval > best_eval:
                best_eval = eval
                best_move = move

            alpha = max(alpha, eval)

        return best_move  # This will now return a tuple (row, col)

    def train(self, num_games=1000):
        wins = 0
        for _ in range(num_games):
            self.board = self.initialize_board()
            self.current_player = 1

            while self.get_moves():
                move = self.get_best_move() if self.current_player == 1 else self.get_random_move()
                self.make_move(move)

                if self.check_win():
                    if self.current_player == -1:
                        wins += 1
                    break

        return wins / num_games

    def get_random_move(self):
        moves = self.get_moves()
        return random.choice(moves)

    def check_win(self):
        # Implement win condition check
        # This is a placeholder and should be implemented based on your game rules
        return False

    def get_board(self):
        return self.board

    def get_current_player(self):
        return self.current_player

# Initialize the bot
bot = GameBot()

File: src/frontend/gameBot.js
// gameBot.js - Board Game AI Bot

class GameBot {
    constructor(boardSize = 8) {
        this.boardSize = boardSize;
        this.board = this.initializeBoard();
        this.currentPlayer = 1; // 1 for player, -1 for bot
    }

    initializeBoard() {
        // Initialize an empty board
        return Array(this.boardSize).fill().map(() => Array(this.boardSize).fill(0));
    }

    getMoves() {
        // Generate all possible moves
        const moves = [];
        for (let i = 0; i < this.boardSize; i++) {
            for (let j = 0; j < this.boardSize; j++) {
                if (this.board[i][j] === 0) {
                    moves.push({ row: i, col: j });
                }
            }
        }
        return moves;
    }

    makeMove(move) {
        if (this.board[move.row][move.col] === 0) {
            this.board[move.row][move.col] = this.currentPlayer;
            this.currentPlayer *= -1; // Switch player
            return true;
        }
        return false;
    }

    undoMove(move) {
        this.board[move.row][move.col] = 0;
        this.currentPlayer *= -1; // Switch player back
    }

    evaluate() {
        // Simple evaluation: difference in piece count
        let score = 0;
        for (let i = 0; i < this.boardSize; i++) {
            for (let j = 0; j < this.boardSize; j++) {
                score += this.board[i][j];
            }
        }
        return score;
    }

    minimax(depth, alpha, beta, maximizingPlayer) {
        if (depth === 0) {
            return this.evaluate();
        }

        const moves = this.getMoves();

        if (maximizingPlayer) {
            let maxEval = -Infinity;
            for (const move of moves) {
                this.makeMove(move);
                const eval = this.minimax(depth - 1, alpha, beta, false);
                this.undoMove(move);
                maxEval = Math.max(maxEval, eval);
                alpha = Math.max(alpha, eval);
                if (beta <= alpha) break;
            }
            return maxEval;
        } else {
            let minEval = Infinity;
            for (const move of moves) {
                this.makeMove(move);
                const eval = this.minimax(depth - 1, alpha, beta, true);
                this.undoMove(move);
                minEval = Math.min(minEval, eval);
                beta = Math.min(beta, eval);
                if (beta <= alpha) break;
            }
            return minEval;
        }
    }

    getBestMove(depth = 3) {
        let bestMove;
        let bestEval = -Infinity;
        const moves = this.getMoves();

        for (const move of moves) {
            this.makeMove(move);
            const eval = this.minimax(depth - 1, -Infinity, Infinity, false);
            this.undoMove(move);

            if (eval > bestEval) {
                bestEval = eval;
                bestMove = move;
            }
        }

        return bestMove;
    }

    // Training and evaluation methods
    train(numGames = 1000) {
        let wins = 0;
        for (let i = 0; i < numGames; i++) {
            this.board = this.initializeBoard();
            this.currentPlayer = 1;
            
            while (this.getMoves().length > 0) {
                const move = this.currentPlayer === 1 ? this.getBestMove() : this.getRandomMove();
                this.makeMove(move);
                
                if (this.checkWin()) {
                    if (this.currentPlayer === -1) wins++;
                    break;
                }
            }
        }
        return wins / numGames;
    }

    getRandomMove() {
        const moves = this.getMoves();
        return moves[Math.floor(Math.random() * moves.length)];
    }

    checkWin() {
        // Implement win condition check
        // This is a placeholder and should be implemented based on your game rules
        return false;
    }

    // Interface for interacting with the bot
    playGame() {
        while (this.getMoves().length > 0) {
            if (this.currentPlayer === 1) {
                console.log("Current board:");
                this.printBoard();
                const move = this.getBestMove();
                console.log(`Bot's move: Row ${move.row}, Col ${move.col}`);
                this.makeMove(move);
            } else {
                const move = this.getHumanMove();
                this.makeMove(move);
            }

            if (this.checkWin()) {
                console.log(`Player ${this.currentPlayer === 1 ? 2 : 1} wins!`);
                break;
            }
        }
        console.log("Game Over");
    }

    getHumanMove() {
        // In a real implementation, this would get input from the user
        // For now, we'll just return a random move
        return this.getRandomMove();
    }

    printBoard() {
        for (let i = 0; i < this.boardSize; i++) {
            console.log(this.board[i].map(cell => cell === 0 ? '.' : cell === 1 ? 'X' : 'O').join(' '));
        }
    }
}

// Example usage
const bot = new GameBot();
console.log("Training bot...");
const winRate = bot.train(1000);
console.log(`Bot win rate after training: ${winRate * 100}%`);
console.log("Starting game...");
bot.playGame();